Nucleo l432kc usb

aurel32 opened the pull request:
These 3 patches add support for USB on the STM32L432 SoC. Tested on a Nucleo L432KC board with an USB connector soldered to pins D2 and D10.

aurel32 committed:
arch: stm32l4: only enable USB OTG on SoCs supporting it

USB OTG is currently enable for the whole STM32L4 family, while only a
few of them actually support it:

- STM32L475, STM32L476 and STM32L496 have an OTG controller
- STM32L432, STM32L433 and STM32L452 have an USB controller
- STM32L431, STM32L451 and STM32L471 do not have any USB controller

Fix that by moving the DT entry from stm32l4.dtsi to stm32l475.dtsi
and by adding a #ifdef #endif around the corresponding DTS fixup
entries.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I just fixed the too long line issue that caused one test to fail.

ydamigos commented on arch/arm/soc/st_stm32/stm32l4/dts.fixup:
```diff
@@ -74,12 +74,22 @@
+#define CONFIG_USB_BASE_ADDRESS			ST_STM32_USB_40006800_BASE_ADDRESS
+#define CONFIG_USB_IRQ				ST_STM32_USB_40006800_IRQ_USB
+#define CONFIG_USB_IRQ_PRI			ST_STM32_USB_40006800_IRQ_USB_PRIORITY
+#define CONFIG_USB_NUM_BIDIR_ENDPOINTS		ST_STM32_USB_40006800_NUM_BIDIR_ENDPOINTS
```
Please add here:
```
#define CONFIG_USB_NUM_IN_ENDPOINTS		ST_STM32_USB_40006800_NUM_IN_ENDPOINTS
#define CONFIG_USB_NUM_OUT_ENDPOINTS		ST_STM32_USB_40006800_NUM_OUT_ENDPOINTS
```
Someone may override the DT usb node on their project and choose some endpoints to be only OUT/IN.

ydamigos reviewed (changes requested):
LGTM. Just a comment.

@erwango Could you create a ticket in ST regarding PWR_CR2_USV. It looks like a bug in STM32Cube.

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -323,6 +323,22 @@ static int usb_dc_stm32_init(void)
+ * in stm32l4xx_ll_pwr.h checks for for the PWR_CR2_PVME1 bit, which is
+ * only available on Cat. 3 devices. It should directly check for
+ * PWR_CR2_USV instead, like done in stm32l4xx_hal_pwr_ex.c.
+ * This is likely a bug in the HAL, add a workaround until it is fixed.
```
Thanks for reporting this, I've filed a bug in HAL (internal ref 47844).
Do you mind updating directly the LL file? This way the patch would also be work for other versions of L432 and for L442 SoC as well. Please don't forget to update Patch list section in ext/hal/st/stm32cube/stm32l4xx/README.

aurel32 replied:
Thanks @erwango for reporting the bug in HAL. I was not aware it is possible to patch HAL directly, it indeed looks a much better solution. I have done that, simplified the changes in drivers/usb/device/usb_dc_stm32.c, and just pushed the changes.

erwango commented on arch/arm/soc/st_stm32/stm32l4/dts.fixup:
```diff
@@ -74,6 +74,14 @@
 #define FLASH_DEV_BASE_ADDRESS		        ST_STM32L4_FLASH_CONTROLLER_40022000_BASE_ADDRESS
 #define FLASH_DEV_NAME			        ST_STM32L4_FLASH_CONTROLLER_40022000_LABEL
 
+#if defined(CONFIG_SOC_STM32L432XC)
```
While, I don't see another way to propose the 2 variants of USB in fixup file (other than put them back in boards, which is not wanted), I'm not quite satisfied with this solution.
Main point is that if one wants to add a new SOC, he'll have to mess with this, while .fixup files should be as transparent as possible.

Can you put a comment to explain we're adding USB configuration depending on supported USB (USB/OTGFS), and one should add new SoC depending on supported USB?

I hope we should get rid of these .fixup files in near future, but meanwhile we should document any "tricks".

aurel32 replied:
I get your point that it will make support for more SoCs less easy. What about this then:

```
#if defined(ST_STM32_USB_40006800_BASE_ADDRESS)
...
#endif

#if defined(ST_STM32_OTGFS_50000000_BASE_ADDRESS)
...
#endif
```

I have tested it works for a nucleo_l432kc, and compiled tested for the stm32l476g_disco. As long as there is a single USB device on the SoC (which seems to be fine for STM32L4 family), that should work.

(BTW, what about enabling support for USB in stm32l476g_disco.dts and stm32l496g_disco.dts. That would make this kind of changes easier to test)

erwango replied:
Sounds like a nice alternative to me. Thanks

erwango replied:
> (BTW, what about enabling support for USB in stm32l476g_disco.dts and stm32l496g_disco.dts. That would make this kind of changes easier to test)

Yeah sure. This should be done as part of #7151, but it is welcome to start now

erwango reviewed (changes requested):
One remaining comment, otherwise LGTM

aurel32 committed:
arch: stm32l432: add support for USB controller

The STM32L432 SoC has a standard non-OTG USB controller. Add an entry
for it in stm32l432.dtsi and add the corresponding DTS fixup entries.
The controller is kept disabled and should be enabled at the board
level.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
ext: hal: stm32cube: fix stm32l4xx VDDUSB supply control

The STM32L4x2 SoCs need to control the isolation of the USB features
from VDDUSB. This is done through the PWR_CR2 bit USV. The STM32L4 HAL
in stm32l4xx_ll_pwr.h wrongly checks for the PWR_CR2_PVME1 bit, which
is only available on Cat. 3 devices. Replace the check by PWR_CR2_USV
like it is already done in stm32l4xx_hal_pwr_ex.c.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: usb_dc_stm32: enable VDDUSB on STM32L4x2

The STM32L4x2 SoCs need to control the isolation of the USB features
from VDDUSB. This is done through the PWR_CR2 bit USV, however the
current code checks for the PWR_CR2_PVME1 bit instead, which is only
available on Cat. 3 devices. This bug is also present int the HAL and
likely copied from there.

Replace the check by PWR_CR2_USV instead.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
@erwango I have just pushed a new version of the patches with the changes discussed earlier today.
